experimental/parser/parse_def.go

```diff
@@ -0,0 +1,363 @@
+
+		switch {
+		case idx < lastFollower:
+			// TODO: if we have already seen a follower at idx, we should
+			// suggest removing this follower. Otherwise, we should suggest
+			// moving this follower before the previous one.
+
+			f := defFollowers[lastFollower]
+			p.Error(errUnexpected{
+				what:  next,
+				where: f.what(p).After(),
+				prev:  f.prev(p),
+				got:   defFollowers[idx].what(p),
+			})
+		case idx == lastFollower:
+			if _, ok := defFollowers[idx].(defBody); ok {
+				// We *do not* pop or diagnose an extra {}; we want this to
+				// be parsed as a loose DeclBody, instead, so we stop
+				// parsing here.
+				break followers
+			}
+
+			f := defFollowers[lastFollower]
+			p.Error(errMoreThanOne{
+				first:  f.prev(p),
+				second: next,
+				what:   f.what(p),
+			})
+		default:
+			lastFollower = idx
+		}
+	}
+
+	// If we didn't see any braces, this def needs to be ended by a semicolon.
+	if p.braces.Nil() {
+		semi, err := p.Punct(p.c, ";", taxa.Def.After())
+		p.args.Semicolon = semi
+		if err != nil {
+			p.Error(err)
+		}
+	}
+
+	// Convert something that looks like an enum value into one.
+	if tyPath := p.args.Type.AsPath(); !tyPath.Nil() && p.args.Name.Nil() &&
+		// The reason for this is because if the user writes `message {}`, we want
+		// to *not* turn it into an enum value with a body.
+		//
+		// TODO: Add a case for making sure that `rpc(foo)` is properly
+		// diagnosed as an anonymous method.
+		p.braces.Nil() {
+		p.args.Name = tyPath.Path
+		p.args.Type = ast.TypeAny{}
+	}
+
+	def := p.NewDeclDef(p.args)
+
+	if !p.inputs.Nil() {
+		parseTypeList(p.parser, p.inputs, def.WithSignature().Inputs(), taxa.MethodIns)
+	}
+	if !p.outputs.Nil() {
+		parseTypeList(p.parser, p.outputs, def.WithSignature().Outputs(), taxa.MethodOuts)
+	} else if !p.outputTy.Nil() {
+		p.Errorf("missing `(...)` around method return type").With(
+			report.Snippetf(p.outputTy, "help: replace this with `(%s)`", p.outputTy.Span().Text()),
+		)
+		def.WithSignature().Outputs().Append(p.outputTy)
+	}
+
+	if !p.braces.Nil() {
+		var in taxa.Noun
+		switch p.kw.Text() {
+		case "message":
+			in = taxa.Message
+		case "enum":
+			in = taxa.Enum
+		case "service":
+			in = taxa.Service
+		case "extend":
+			in = taxa.Extend
+		case "group":
+			in = taxa.Field
+		case "oneof":
+			in = taxa.Oneof
+		case "rpc":
+			in = taxa.Method
+		default:
+			in = taxa.Def
+		}
+
+		def.SetBody(parseBody(p.parser, p.braces, in))
+	}
+
+	return def
+}
+
+type defInputs struct{}
+
+func (defInputs) what(*defParser) taxa.Noun  { return taxa.MethodIns }
+func (defInputs) canStart(p *defParser) bool { return p.c.Peek().Text() == "(" }
+
+func (defInputs) parse(p *defParser) report.Span {
+	next := p.c.Pop()
+	if next.IsLeaf() {
+		return report.Span{} // Diagnosed by the lexer.
+	}
+
+	if p.inputs.Nil() {
+		p.inputs = next
+	}
+	return next.Span()
+}
+
+func (defInputs) prev(p *defParser) report.Span {
+	if p.inputs.Nil() {
+		return report.Span{}
+	}
```

Comment:
Currently that may panic. I could leave a TODO to make sure that no call to Span() panics... it would be a big yak shave so I would rather do it in its own PR in one shot. Maybe I'll do it today...

Reply:
See https://github.com/bufbuild/protocompile/pull/395